builtin/providers/aws/resource_aws_instance.go

```diff
@@ -311,24 +345,27 @@ func resourceAwsInstanceCreate(d *schema.ResourceData, meta interface{}) error {
 	}
 
 	if len(blockDevices) > 0 {
-		runOpts.BlockDevices = make([]ec2.BlockDeviceMapping, len(blockDevices))
+		runOpts.BlockDeviceMappings = make([]ec2.BlockDeviceMapping, len(blockDevices))
 		for i, v := range blockDevices {
 			bd := v.(map[string]interface{})
-			runOpts.BlockDevices[i].DeviceName = bd["device_name"].(string)
-			runOpts.BlockDevices[i].VolumeType = bd["volume_type"].(string)
-			runOpts.BlockDevices[i].VolumeSize = int64(bd["volume_size"].(int))
-			runOpts.BlockDevices[i].DeleteOnTermination = bd["delete_on_termination"].(bool)
+			runOpts.BlockDeviceMappings[i].DeviceName = aws.String(bd["device_name"].(string))
+			runOpts.BlockDeviceMappings[i].EBS = &ec2.EBSBlockDevice{
+				VolumeType:          aws.String(bd["volume_type"].(string)),
+				VolumeSize:          aws.Integer(bd["volume_size"].(int)),
+				DeleteOnTermination: aws.Boolean(bd["delete_on_termination"].(bool)),
+			}
+
 			if v, ok := bd["virtual_name"].(string); ok {
-				runOpts.BlockDevices[i].VirtualName = v
+				runOpts.BlockDeviceMappings[i].VirtualName = aws.String(v)
 			}
-			if v, ok := bd["snapshot_id"].(string); ok {
-				runOpts.BlockDevices[i].SnapshotId = v
+			if v, ok := bd["snapshot_id"].(string); ok && v != "" {
+				runOpts.BlockDeviceMappings[i].EBS.SnapshotID = aws.String(v)
 			}
 			if v, ok := bd["encrypted"].(bool); ok {
-				runOpts.BlockDevices[i].Encrypted = v
+				runOpts.BlockDeviceMappings[i].EBS.Encrypted = aws.Boolean(v)
 			}
-			if v, ok := bd["iops"].(int); ok {
-				runOpts.BlockDevices[i].IOPS = int64(v)
+			if v, ok := bd["iops"].(int); ok && v > 0 {
+				runOpts.BlockDeviceMappings[i].EBS.IOPS = aws.Integer(v)
 			}
```

Comment:
So there's a bit of weirdness this'll introduce around ephemeral devices - people could kinda sorta hack'em in with the old way. But I don't think it's a problem because #1045 will redo all of that anyways.